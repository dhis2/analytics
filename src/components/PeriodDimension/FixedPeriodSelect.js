import { SingleSelectField, SingleSelectOption } from '@dhis2/ui'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import i18n from '../../locales/index.js'
import FixedPeriodFilter from './FixedPeriodFilter.js'
import styles from './styles/FixedPeriodSelect.style.js'
import {
    parsePeriodCode,
    getFixedPeriodsOptionsById as getPeriodById,
    getYearOffsetFromNow,
} from './utils/fixedPeriods.js'

class FixedPeriodSelect extends Component {
    state = {
        periodType: '',
        year: '',
        options: null,
    }

    static getDerivedStateFromProps({ value, allowedPeriodTypes }) {
        const period = parsePeriodCode(value, allowedPeriodTypes)

        if (!period) {
            return null
        }

        return {
            periodType: period.id,
            year: period.year,
            options: period.options,
        }
    }

    onSelectPeriodType = (periodType) => {
        this.setState({
            periodType,
            options: this.getUpdatedOptions(periodType, this.state.year),
        })
        this.props.onChange()
    }

    onSelectYear = (year) => {
        this.setState({
            year,
            options: this.getUpdatedOptions(this.state.periodType, year),
        })
        this.props.onChange()
    }

    onSelectPeriod = ({ selected: periodId }, event) => {
        const period = this.state.options.find(({ id }) => id === periodId)
        this.props.onChange(period, event)
    }

    getUpdatedOptions(periodType, year) {
        const offset = getYearOffsetFromNow(year)
        return getPeriodById(periodType).getPeriods({ offset })
    }

    render() {
        return (
            <div
                data-test={this.props.dataTest}
                className={this.props.className}
            >
                <div className="row">
                    <FixedPeriodFilter
                        allowedPeriodTypes={this.props.allowedPeriodTypes}
                        currentPeriodType={this.state.periodType}
                        currentYear={this.state.year}
                        onSelectPeriodType={this.onSelectPeriodType}
                        onSelectYear={this.onSelectYear}
                    />
                </div>

                {this.state.periodType && this.state.year && (
                    <SingleSelectField
                        label={i18n.t('Period')}
                        onChange={this.onSelectPeriod}
                        dense
                        selected={this.props.value}
                    >
                        {this.state.options.map((option) => (
                            <SingleSelectOption
                                key={option.id}
                                value={option.id}
                                label={option.name}
                            />
                        ))}
                    </SingleSelectField>
                )}
                <style jsx>{styles}</style>
            </div>
        )
    }
}

FixedPeriodSelect.defaultProps = {
    dataTest: 'dhis2-analytics-fixedperiodselect',
    value: '',
}

FixedPeriodSelect.propTypes = {
    onChange: PropTypes.func.isRequired,
    allowedPeriodTypes: PropTypes.arrayOf(PropTypes.string),
    className: PropTypes.string,
    dataTest: PropTypes.string,
    value: PropTypes.string,
}

export default FixedPeriodSelect
