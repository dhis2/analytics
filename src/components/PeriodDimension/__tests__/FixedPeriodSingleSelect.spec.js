import React from 'react'
import { shallow } from 'enzyme'
import FixedPeriodSelect from '../FixedPeriodSelect.js'

describe('The Fixed Period Single Select component', () => {
    let props
    let shallowFixedPeriodSelect

    const getWrapper = () => {
        if (!shallowFixedPeriodSelect) {
            shallowFixedPeriodSelect = shallow(<FixedPeriodSelect {...props} />)
        }
        return shallowFixedPeriodSelect
    }

    beforeEach(() => {
        props = {
            value: '201405',
            onChange: () => {},
        }
        shallowFixedPeriodSelect = undefined
    })

    it('matches the snapshot', () => {
        const wrapper = getWrapper()

        expect(wrapper).toMatchSnapshot()
    })
})
