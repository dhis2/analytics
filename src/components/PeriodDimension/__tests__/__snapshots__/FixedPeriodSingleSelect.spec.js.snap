// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Fixed Period Single Select component matches the snapshot 1`] = `
<div
  data-test="dhis2-analytics-fixedperiodselect"
>
  <div
    className="row"
  >
    <FixedPeriodFilter
      currentPeriodType="MONTHLY"
      currentYear="2014"
      excludedPeriodTypes={Array []}
      onSelectPeriodType={[Function]}
      onSelectYear={[Function]}
    />
  </div>
  <SingleSelectField
    clearText={[Function]}
    dataTest="dhis2-uiwidgets-singleselectfield"
    dense={true}
    empty={[Function]}
    filterPlaceholder={[Function]}
    label="Period"
    loadingText={[Function]}
    noMatchText={[Function]}
    onChange={[Function]}
    selected="201405"
  >
    <SingleSelectOption
      dataTest="dhis2-uicore-singleselectoption"
      key="201401"
      label="January 2014"
      value="201401"
    />
    <SingleSelectOption
      dataTest="dhis2-uicore-singleselectoption"
      key="201402"
      label="February 2014"
      value="201402"
    />
    <SingleSelectOption
      dataTest="dhis2-uicore-singleselectoption"
      key="201403"
      label="March 2014"
      value="201403"
    />
    <SingleSelectOption
      dataTest="dhis2-uicore-singleselectoption"
      key="201404"
      label="April 2014"
      value="201404"
    />
    <SingleSelectOption
      dataTest="dhis2-uicore-singleselectoption"
      key="201405"
      label="May 2014"
      value="201405"
    />
    <SingleSelectOption
      dataTest="dhis2-uicore-singleselectoption"
      key="201406"
      label="June 2014"
      value="201406"
    />
    <SingleSelectOption
      dataTest="dhis2-uicore-singleselectoption"
      key="201407"
      label="July 2014"
      value="201407"
    />
    <SingleSelectOption
      dataTest="dhis2-uicore-singleselectoption"
      key="201408"
      label="August 2014"
      value="201408"
    />
    <SingleSelectOption
      dataTest="dhis2-uicore-singleselectoption"
      key="201409"
      label="September 2014"
      value="201409"
    />
    <SingleSelectOption
      dataTest="dhis2-uicore-singleselectoption"
      key="201410"
      label="October 2014"
      value="201410"
    />
    <SingleSelectOption
      dataTest="dhis2-uicore-singleselectoption"
      key="201411"
      label="November 2014"
      value="201411"
    />
    <SingleSelectOption
      dataTest="dhis2-uicore-singleselectoption"
      key="201412"
      label="December 2014"
      value="201412"
    />
  </SingleSelectField>
  <style />
</div>
`;
